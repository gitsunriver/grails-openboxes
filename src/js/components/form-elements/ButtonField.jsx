import React from 'react';
import PropTypes from 'prop-types';

const ButtonField = (props) => {
  const {
    fieldName,
    fieldConfig: { buttonLabel: ButtonLabel, getDynamicAttr, attributes = {} },
    fieldPreview,
  } = props;
  const dynamicAttr = getDynamicAttr ? getDynamicAttr(props) : {};
  const attr = { disabled: fieldPreview, ...attributes, ...dynamicAttr };

  return (
    <button type="button" key={fieldName} className="btn btn-outline-primary" {...attr} >
      {
        typeof ButtonLabel === 'string' ? ButtonLabel : <ButtonLabel />
      }
    </button>
  );
};

export default ButtonField;

ButtonField.propTypes = {
  fieldName: PropTypes.string.isRequired,
  fieldConfig: PropTypes.shape({
    getDynamicAttr: PropTypes.func,
  }).isRequired,
  fieldPreview: PropTypes.bool,
};

ButtonField.defaultProps = {
  fieldPreview: false,
};
