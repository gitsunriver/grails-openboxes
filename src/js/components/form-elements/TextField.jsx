import React from 'react';

import BaseField from './BaseField';
import Input from '../../utils/Input';

const TextField = (props) => {
  const renderInput = attributes => (
    <Input
      className="form-control"
      {...attributes}
    />
  );

  return (
    <BaseField
      {...props}
      renderInput={renderInput}
    />
  );
};

export default TextField;
