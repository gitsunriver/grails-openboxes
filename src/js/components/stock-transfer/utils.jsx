import _ from 'lodash';

const PENDING = 'PENDING';
const CANCELED = 'CANCELED';

function extractItem(item, status) {
  const { destinationBinLocation, destinationZone } = item;
  return {
    ...item,
    destinationBinLocation: {
      id: destinationBinLocation && destinationBinLocation.id ? destinationBinLocation.id : '',
      name: destinationBinLocation && destinationBinLocation.name ? destinationBinLocation.name : '',
      zoneId: destinationZone && destinationZone.id ? destinationZone.id : null,
      zoneName: destinationZone && destinationZone.name ? destinationZone.name : null,
    },
    quantity: status === PENDING ? '' : item.quantity,
  };
}

function extractSplitItem(item, splitItem, status) {
  return {
    ...splitItem,
    destinationBinLocation: {
      id: splitItem.destinationBinLocation.id ? splitItem.destinationBinLocation.id : '',
      name: splitItem.destinationBinLocation.name ? splitItem.destinationBinLocation.name : '',
      zoneId: splitItem.destinationBinLocation.zone ?
        splitItem.destinationBinLocation.zone.id : null,
      zoneName: splitItem.destinationBinLocation.zone ?
        splitItem.destinationBinLocation.zone.name : null,
    },
    quantity: status === PENDING ? '' : item.quantity,
    quantityOnHand: item.quantityOnHand,
    referenceId: item.id, // set a referenceId from original item
  };
}

export function extractStockTransferItems(stockTransfer) {
  const stockTransferItems = [];
  _.forEach(stockTransfer.stockTransferItems, (item) => {
    stockTransferItems.push(extractItem(item, stockTransfer.status));
    if (item.splitItems.length > 0) {
      _.forEach(item.splitItems, (splitItem) => {
        stockTransferItems.push(extractSplitItem(item, splitItem, stockTransfer.status));
      });
    }
  });
  return stockTransferItems;
}

export function extractNonCanceledItems(stockTransfer) {
  const stockTransferItems = [];
  _.forEach(stockTransfer.stockTransferItems, (item) => {
    if (item.status !== CANCELED) {
      stockTransferItems.push(extractItem(item, stockTransfer.status));
    }
    if (item.splitItems.length > 0) {
      _.forEach(item.splitItems, (splitItem) => {
        stockTransferItems.push(extractSplitItem(item, splitItem, stockTransfer.status));
      });
    }
  });
  return stockTransferItems;
}

export function prepareRequest(stockTransfer, status) {
  const originalItems = _.filter(stockTransfer.stockTransferItems, item => !item.referenceId);
  const stockTransferItems = _.map(originalItems, item => ({
    ...item,
    splitItems: _.filter(
      stockTransfer.stockTransferItems,
      splitItem => splitItem.referenceId === item.id,
    ),
    quantity: item.quantity === '' ? item.quantityOnHand : item.quantity,
  }));

  return { ...stockTransfer, stockTransferItems, status };
}
