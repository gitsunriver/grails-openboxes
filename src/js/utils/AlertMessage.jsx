import React from 'react';
import PropTypes from 'prop-types';

const AlertMessage = ({
  show, message, danger, close,
}) => {
  if (show) {
    return (
      <div
        className={`alert ${danger ? 'alert-danger' : 'alert-warning'}`}
        style={{ cursor: close ? 'pointer' : 'default' }}
        role="presentation"
        onClick={() => close()}
      >
        <i className={`fa ${danger ? 'fa-times-circle' : 'fa-exclamation-circle'} pr-2`} />
        {message}
      </div>);
  }

  return null;
};

export default AlertMessage;

AlertMessage.propTypes = {
  show: PropTypes.bool,
  message: PropTypes.string,
  danger: PropTypes.bool,
  close: PropTypes.func,
};

AlertMessage.defaultProps = {
  show: false,
  message: '',
  danger: false,
  close: () => null,
};
