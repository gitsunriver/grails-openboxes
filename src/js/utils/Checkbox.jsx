import React from 'react';
import PropTypes from 'prop-types';

const Checkbox = ({
  value, indeterminate, custom, ...props
}) => {
  const onChange = (event) => {
    const { checked } = event.target;

    if (props.onChange) {
      props.onChange(checked);
    }
  };

  if (custom) {
    return (
      <div className="custom-checkbox">
        <input
          type="checkbox"
          checked={value}
          {...props}
          onChange={onChange}
        />
        <label htmlFor={props.id} />
      </div>
    );
  }

  return (
    <input
      type="checkbox"
      // eslint-disable-next-line no-param-reassign,no-return-assign
      ref={elem => elem && (elem.indeterminate = indeterminate)}
      checked={value}
      {...props}
      onChange={onChange}
    />);
};

export default Checkbox;

Checkbox.propTypes = {
  onChange: PropTypes.func,
  value: PropTypes.bool,
  indeterminate: PropTypes.bool,
  custom: PropTypes.bool,
  id: PropTypes.string,
};

Checkbox.defaultProps = {
  onChange: null,
  value: null,
  indeterminate: false,
  custom: false,
  id: '',
};
