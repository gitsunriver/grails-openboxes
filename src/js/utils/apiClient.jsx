/* eslint-disable no-console */
import axios from 'axios';

const justRejectRequestError = error => Promise.reject(error);

const apiClient = axios.create({});

const handleSuccess = response => response;

const handleError = (error) => {
  switch (error.response.status) {
    case 400:
      console.error('Bad request.');
      break;
    case 401:
      console.error('Unauthorized.');
      break;
    case 403:
      console.error('Access denied.');
      break;
    case 404:
      console.error('Not found.');
      break;
    case 500:
      console.error('Internal server error.');
      break;
    default: {
      console.error(error);
    }
  }
  return Promise.reject(error);
};

apiClient.interceptors.response.use(handleSuccess, handleError);
apiClient.interceptors.request.use(config => config, justRejectRequestError);

export default apiClient;
