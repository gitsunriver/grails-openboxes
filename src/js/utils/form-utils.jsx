import React from 'react';
import PropTypes from 'prop-types';

export const renderFormField = (fieldConfig, fieldName, props = {}) => {
  const FieldType = fieldConfig.type;

  return (
    <FieldType
      key={fieldName}
      fieldName={fieldName}
      fieldConfig={fieldConfig}
      {...props}
    />
  );
};

export const renderField = ({
  renderInput,
  attributes: { required, hidden, ...otherAttributes },
  label: FieldLabel,
  touched: fieldTouched,
  input,
  meta: { touched, error },
}) => {
  const attr = { id: input.name, ...otherAttributes };

  const className = `form-group ${required ? 'required' : ''} ${hidden ? 'd-none' : ''} ${(touched || fieldTouched) && error ? 'has-error' : ''}`;

  return (
    <div className={`padding-left-md padding-right-md ${className}`}>
      <div className="row">
        {
          typeof FieldLabel === 'string' ?
            <label htmlFor={attr.id} className="col-md-2 col-form-label">{ FieldLabel }</label> :
            <FieldLabel />
        }
        {renderInput(input, attr)}
      </div>
      <div className="row">
        <div className="col-md-2" />
        <div className="help-block col-md-4" style={{ float: 'left' }}>
          { touched || fieldTouched ? error : '' }
        </div>
      </div>
    </div>
  );
};

renderField.propTypes = {
  renderInput: PropTypes.func.isRequired,
  attributes: PropTypes.shape({}).isRequired,
  label: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.func,
  ]),
  touched: PropTypes.bool,
  input: PropTypes.shape({}).isRequired,
  meta: PropTypes.shape({}).isRequired,
};

renderField.defaultProps = {
  touched: false,
  label: '',
};
