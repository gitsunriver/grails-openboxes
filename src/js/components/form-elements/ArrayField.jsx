import _ from 'lodash';
import React from 'react';
import PropTypes from 'prop-types';
import { FieldArray } from 'redux-form';

import { renderFormField } from '../../utils/form-utils';

const ArrayField = (props) => {
  const { fieldName, fieldConfig, ...otherProps } = props;

  const renderFieldArray = ({ fieldsConfig, properties, fields }) => {
    const AddButton = fieldsConfig.addButton;
    const addRow = (row = {}) => fields.push(row);

    return (
      <div>
        <table className="table table-striped">
          <thead>
            <tr>
              { _.map(fieldsConfig.fields, (config, name) =>
                <th key={name}>{config.label}</th>) }
            </tr>
          </thead>
          <tbody>
            {fields.map((field, index) => (
            // eslint-disable-next-line react/no-array-index-key
              <tr key={index}>
                { _.map(fieldsConfig.fields, (config, name) => (
                  <td key={`${field}.${name}`}>
                    { renderFormField(config, `${field}.${name}`, {
                      ...properties,
                      arrayField: true,
                      addRow,
                      removeRow: () => fields.remove(index),
                    })}
                  </td>
              )) }
              </tr>))}
          </tbody>
        </table>
        <div className="text-center">
          {
            typeof AddButton === 'string' ?
              <button type="button" className="btn btn-outline-success margin-bottom-lg" onClick={() => addRow()}>
                {AddButton}
              </button>
              : <AddButton addRow={addRow} />
          }
        </div>
      </div>
    );
  };

  renderFieldArray.propTypes = {
    fieldsConfig: PropTypes.shape({}).isRequired,
    fields: PropTypes.oneOfType([
      PropTypes.shape({}),
      PropTypes.arrayOf(PropTypes.shape({})),
    ]).isRequired,
    properties: PropTypes.shape({}),
  };

  renderFieldArray.defaultProps = {
    properties: {},
  };

  return (
    <FieldArray
      key={fieldName}
      name={fieldName}
      component={renderFieldArray}
      fieldsConfig={fieldConfig}
      properties={otherProps}
    />
  );
};

export default ArrayField;

ArrayField.propTypes = {
  fieldName: PropTypes.string.isRequired,
  fieldConfig: PropTypes.shape({
    getDynamicAttr: PropTypes.func,
    fields: PropTypes.shape({}),
  }).isRequired,
};
