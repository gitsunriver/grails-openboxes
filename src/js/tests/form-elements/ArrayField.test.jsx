import React from 'react';
import renderer from 'react-test-renderer';
import ArrayField from '../../components/form-elements/ArrayField';
import TextField from '../../components/form-elements/TextField';
import ButtonField from '../../components/form-elements/ButtonField';
import { renderFormField } from '../../utils/form-utils';

jest.mock('redux-form', () => ({
  Field: (props) => {
    const { component: Component, name, ...others } = props;
    const input = { onChange: () => {}, name };
    const meta = {};
    return <Component input={input} meta={meta} {...others} />;
  },
  FieldArray: (props) => {
    const { component: Component, name, ...others } = props;
    const fields = { map: callback => [`${name}[0]`, `${name}[1]`].map(callback) };
    return <Component fields={fields} {...others} />;
  },
  formValueSelector: () => {},
}));

describe('ArrayField component is correctly rendering', () => {
  it('with Add button', () => {
    const fieldConfig = {
      type: ArrayField,
      addButton: 'Add item',
      fields: {
        textField: {
          type: TextField,
          label: 'Name',
        },
        textField2: {
          type: TextField,
          label: 'Name',
        },
        button: {
          type: ButtonField,
          buttonLabel: 'Delete',
          getDynamicAttr: ({ removeRow }) => ({
            onClick: removeRow,
          }),
          attributes: {
            className: 'btn btn-outline-danger',
          },
        },
      },
    };

    const rendered = renderer.create(renderFormField(fieldConfig, 'test-field'));

    expect(rendered.toJSON()).toMatchSnapshot();
  });

  it('with no Add button', () => {
    const fieldConfig = {
      type: ArrayField,
      fields: {
        textField: {
          type: TextField,
          label: 'Name',
        },
        textField2: {
          type: TextField,
          label: 'Name',
        },
        button: {
          type: ButtonField,
          buttonLabel: 'Delete',
          getDynamicAttr: ({ removeRow }) => ({
            onClick: removeRow,
          }),
          attributes: {
            className: 'btn btn-outline-danger',
          },
        },
      },
    };

    const rendered = renderer.create(renderFormField(fieldConfig, 'test-field'));

    expect(rendered.toJSON()).toMatchSnapshot();
  });
});

