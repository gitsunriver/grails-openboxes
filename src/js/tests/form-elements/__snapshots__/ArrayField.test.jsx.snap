// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArrayField component is correctly rendering with Add button 1`] = `
<div>
  <table
    className="table table-striped text-center border"
  >
    <thead>
      <tr>
        <th>
          Name
        </th>
        <th>
          Name
        </th>
        <th />
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          className="align-middle"
        >
          <div
            className="form-group my-0   "
          >
            <input
              className="form-control"
              id="test-field[0].textField"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value={undefined}
            />
          </div>
        </td>
        <td
          className="align-middle"
        >
          <div
            className="form-group my-0   "
          >
            <input
              className="form-control"
              id="test-field[0].textField2"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value={undefined}
            />
          </div>
        </td>
        <td
          className="align-middle"
        >
          <button
            className="btn btn-outline-danger"
            onClick={[Function]}
            type="button"
          >
            Delete
          </button>
        </td>
      </tr>
      <tr>
        <td
          className="align-middle"
        >
          <div
            className="form-group my-0   "
          >
            <input
              className="form-control"
              id="test-field[1].textField"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value={undefined}
            />
          </div>
        </td>
        <td
          className="align-middle"
        >
          <div
            className="form-group my-0   "
          >
            <input
              className="form-control"
              id="test-field[1].textField2"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value={undefined}
            />
          </div>
        </td>
        <td
          className="align-middle"
        >
          <button
            className="btn btn-outline-danger"
            onClick={[Function]}
            type="button"
          >
            Delete
          </button>
        </td>
      </tr>
    </tbody>
  </table>
  <div
    className="text-center"
  >
    <button
      className="btn btn-outline-success margin-bottom-lg"
      onClick={[Function]}
      type="button"
    >
      Add item
    </button>
  </div>
</div>
`;

exports[`ArrayField component is correctly rendering with no Add button 1`] = `
<div>
  <table
    className="table table-striped text-center border"
  >
    <thead>
      <tr>
        <th>
          Name
        </th>
        <th>
          Name
        </th>
        <th />
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          className="align-middle"
        >
          <div
            className="form-group my-0   "
          >
            <input
              className="form-control"
              id="test-field[0].textField"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value={undefined}
            />
          </div>
        </td>
        <td
          className="align-middle"
        >
          <div
            className="form-group my-0   "
          >
            <input
              className="form-control"
              id="test-field[0].textField2"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value={undefined}
            />
          </div>
        </td>
        <td
          className="align-middle"
        >
          <button
            className="btn btn-outline-danger"
            onClick={[Function]}
            type="button"
          >
            Delete
          </button>
        </td>
      </tr>
      <tr>
        <td
          className="align-middle"
        >
          <div
            className="form-group my-0   "
          >
            <input
              className="form-control"
              id="test-field[1].textField"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value={undefined}
            />
          </div>
        </td>
        <td
          className="align-middle"
        >
          <div
            className="form-group my-0   "
          >
            <input
              className="form-control"
              id="test-field[1].textField2"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value={undefined}
            />
          </div>
        </td>
        <td
          className="align-middle"
        >
          <button
            className="btn btn-outline-danger"
            onClick={[Function]}
            type="button"
          >
            Delete
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
